### ISSUE TITLE
[TySan] False negative for int/short aliasing

### ISSUE THREAD

**Issue by boomanaiden154** [llvm/llvm-project#206226]
```c
#define a() 0
int b, c, d;
short i;
int f1(int j) {
  for (; 0 < 1;) {
  f:
    if (d == 0)
      return 0;
    if (c > 7)
      goto f;
    if (j > 0)
      return 0;
    if (!c) {
      for (int g; g < 1;)
        if (a() != 0)
          return 0;
      for (int h; h < 1; h++)
        if (b == 0)
          return 0;
    }
  }
  return 1;
}
void aa(int j, int *k) {
  short *a = (short *)k;
  for (; (f1(j + 9) < a[0]) + j;) {
    for (int e = 0; e < 1; e = 0 == i)
      a = 0;
    a += 4;
  }
}
void aj(int j) {
  int f = 0;
  aa(j, &f);
}
int main() { aj(0); }
```

Taking the following snippet (from #187952), and then trying to run it under TySan:
```shell
clang -O2 small.c -o test -fsanitize=type && ./test
```

Does not diagnose anything. Presumably we should be diagnosing something though given in `aj` we declare an `int`, `f`, pass it by pointer to `aa`, and then try and access it through a `short *`.

**Comment by boomanaiden154** [comment]
CC @jyknight @fhahn

**Comment by boomanaiden154** [comment]
Nevermind, looks like this is actually caught at `-O0`:
```
==440251==ERROR: TypeSanitizer: type-aliasing-violation on address 0x7fff681112f8 (pc 0x55c6473aee07 bp 0x7fff68111270 sp 0x7fff681111f8 tid 440251)
READ of size 2 at 0x7fff681112f8 with type short accesses an existing object of type int
    #0 0x55c6473aee06 in aa (/tmp/test1+0x35e06)
```

Probably the same issue as #120855.

**Comment by gbMattN** [comment]
The [tysan docs](https://clang.llvm.org/docs/TypeSanitizer.html) mention using `-O1` at most due to optimizations possibly (incorrectly) removing the code that causes the violation... from the looks of it maybe the recommendation should be `-O0`...
